### PR TITLE
7300 durable Recorder

### PR DESCRIPTION
**Pull request by turadg**
refs: #7300

## Description

Publishing to vstorage in upgradable contracts requires a durable object to keep the storageNode. We explored other designs such as https://github.com/Agoric/agoric-sdk/pull/7341 but ran into problems. @gibson042 and I concluded that wrapping the `publisher` was the cleanest way forward.

This wraps only the `publisher` of the PublishKit  so `subscriber` access does not need to await resolution of storageNode.

It also wraps it opaquely because there are semantic differences in its operation and it has different requirements than a regular publisher. This implies that when introducing vstorage writes to an existing PublishKit consumer, downstream changes will be necessary. (E.g. [makeInstallationPublisher](https://github.com/Agoric/agoric-sdk/blob/79bd8b5baef7fe3a4139e3b6982e82715fe12a2b/packages/cosmic-swingset/src/chain-main.js#L282-L285) may need to return a `Recorder` and be renaming accordingly.)

This provides a design for solving #7300 and exercises it with Vault Factory. A few other places will also need updating so in a separate PR I'll do that and remove `pipeTopicToStorage`.

### Security Considerations

Authorities given to Zoe contract support functions.

### Scaling Considerations

Each durable contract with public topics will require one durable kind for the durable PublishKit and another for the Recorder. Each topic will instantiate them both.

### Documentation Considerations

--

### Testing Considerations

Zoe could use tests for these contract supports, though I contend that "real world" testing in Inter Protocol is sufficient for now.

### CODE REVIEW

File: packages/inter-protocol/src/vaultFactory/vaultDirector.js [hunk 1]
```diff
@@ -115,16 +111,15 @@ export const prepareVaultDirector = (
...
   const vaultParamManagers = makeScalarMapStore('vaultParamManagers');
 
-  /** @type {PublishKit<MetricsNotification>} */
-  const { publisher: metricsPublisher, subscriber: metricsSubscriber } =
-    makeVaultDirectorPublishKit();
-
   const metricsNode = E(storageNode).makeChildNode('metrics');
-  pipeTopicToStorage(metricsSubscriber, metricsNode, marshaller);
+  /** @type {import('@agoric/zoe/src/contractSupport/recorder.js').EventualRecorderKit<MetricsNotification>} */
+  const metricsKit = makeRecorderKit.async(metricsNode);
```
Comment (gibson042) [inline packages/inter-protocol/src/vaultFactory/vaultDirector.js]: This invocation surprises me; I think we generally try not to hang properties off of functions. It would be more conventional to either drop this variation altogether or achieve it some other way (e.g., a distinct maker).

File: packages/inter-protocol/src/vaultFactory/vaultFactory.js [hunk 2]
```diff
@@ -103,7 +106,9 @@ export const start = async (zcf, privateArgs, baggage) => {
     timerService,
     auctioneerPublicFacet,
     storageNode,
+    // XXX redundant with makeStoredPublishKit once we excise deprecated kits for governance
```
Comment (gibson042) [inline packages/inter-protocol/src/vaultFactory/vaultFactory.js]:
```suggestion
    // XXX redundant with makeRecorderKit once we excise deprecated kits for governance
```
?

File: packages/inter-protocol/src/vaultFactory/vaultHolder.js [hunk 3]
```diff
@@ -33,19 +26,20 @@ const HolderI = M.interface('holder', {
   makeCloseInvitation: M.call().returns(M.promise()),
   makeTransferInvitation: M.call().returns(M.promise()),
 });
+
+/** @type {{ [name: string]: [ description: string, valueShape: Pattern ] }} */
+const PUBLIC_TOPICS = {
+  vault: ['Vault holder status', M.any()],
```
Comment (gibson042) [inline packages/inter-protocol/src/vaultFactory/vaultHolder.js:31]: Is the second element of this array intentionally unused? I was trying to suggest a better name for PUBLIC_TOPICS, but can't actually determine what that would be.

File: packages/zoe/src/contractSupport/recorder.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
+import { prepareDurablePublishKit } from '@agoric/notifier';
+import { M, prepareExoClass } from '@agoric/vat-data';
+import { E } from '@endo/eventual-send';
+
+/**
+ * Recorders support publishing data to vstorage.
+ *
+ * `Recorder` is similar to `Publisher` (in that they send out data) but has different signatures:
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js:14]: I like the name!

File: packages/zoe/src/contractSupport/recorder.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+ * - methods are async because they await remote calls to Marshaller and StorageNode
+ * - method names more aligned with [async iterators](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-iterator-interface)
+ * - omits fail()
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js:17]:
```suggestion
 * - omits fail()/throw()
```

File: packages/zoe/src/contractSupport/recorder.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
+import { prepareDurablePublishKit } from '@agoric/notifier';
+import { M, prepareExoClass } from '@agoric/vat-data';
+import { E } from '@endo/eventual-send';
+
+/**
+ * Recorders support publishing data to vstorage.
+ *
+ * `Recorder` is similar to `Publisher` (in that they send out data) but has different signatures:
+ * - methods are async because they await remote calls to Marshaller and StorageNode
+ * - method names more aligned with [async iterators](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-iterator-interface)
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js]: These names I'm less happy with, though... the async iterator interface describes objects that _supply_ data on demand—`next` is a request for the next result (and is almost always called without an argument), and `return` and `throw` are used to inform such a supplier that no further `next` requests should be expected. The interface seems like a poor fit for an object whose methods are intended to _accept_ data for constructing such a sequence (although a good fit for an the `subscriber` facet of a publish kit).

File: packages/zoe/src/contractSupport/recorder.js [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  };
+  /**
+   * @param {ERef<StorageNode>} nodeP
+   * @returns {EventualRecorderKit<any>}
+   */
+  makeRecorderKit.async = nodeP => {
+    const dpk = makeDurablePublishKit();
+    return harden({
+      ...dpk,
+      recorder: E.when(nodeP, node => makeRecorder(dpk.publisher, node)),
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js]: This doesn't feel quite right; probably there should instead be a single conventional `makeRecorderKit` with a signature like `(storageNode: StorageNode) => RecorderKit`, and than another independent function (i.e., defined external to `prepareRecorderKit`) with one of the following signatures (the first being preferable if possible, the second being an acceptable compromise if necessary):
1. `(storageNodeP: ERef<StorageNode>) => Promise<RecorderKit>`
2. `(storageNodeP: ERef<StorageNode>) => { subscriber: Pick<RecorderKit, 'subscriber'>, recorderP: Promise<Pick<RecorderKit, 'recorder'>>}`

And if the second _is_ necessary, then I would expect packages/zoe/src/contractSupport/recorder.js to define only one prepare function (`prepareRecorder`) and one **non**-prepare helper function in alignment with [A Taxonomy of Exo-making Functions](https://github.com/endojs/endo/blob/master/packages/exo/docs/exo-taxonomy.md), e.g.
```js
const defineERecorderKit = ({ makeRecorder, makeDurablePublishKit }) => {
  const makeRecorderKitWhen = storageNodeP => {
    const { publisher, subscriber } = makeDurablePublishKit();
    const recorderP = E.when(storageNodeP, storageNode =>
      makeRecorder(publisher, storageNode),
    );
    return { subscriber, recorderP };
  };
  return makeRecorderKitWhen;
};
```
for use by packages/inter-protocol/src/vaultFactory/vaultFactory.js and packages/inter-protocol/test/vaultFactory/vault-contract-wrapper.js to define a `makeRecorderKitWhen`
```diff
-  const makeRecorderKit = prepareRecorderKit(baggage, marshaller);
+  const makeRecorder = prepareRecorder(baggage, marshaller);
+  const makeRecorderKitWhen = defineERecorderKit({
+    makeRecorder,
+    makeDurablePublishKit,
+  });
```
which they would pass down where this PR currently sends `makeRecorderKit` for use in deeper layers like
```diff
-  const assetKit = makeRecorderKit.async(storageNode);
+  const { subscriber: assetSubscriber, recorderP: assetRecorderP } =
+    makeRecorderKitWhen(storageNode);
```

File: packages/inter-protocol/src/vaultFactory/vaultManager.js [hunk 8]
```diff
@@ -245,13 +239,13 @@ export const prepareVaultManagerKit = (
...
-      storageNode,
+      assetKit.subscriber,
+      E(assetKit.recorder).getStorageNode(),
     ),
     metrics: makePublicTopic(
       'Vault Factory metrics',
-      metricsSubscriber,
-      metricsNode,
+      metricsTopicKit.subscriber,
+      E(metricsTopicKit.recorder).getStorageNode(),
```
Comment (gibson042) [inline packages/inter-protocol/src/vaultFactory/vaultManager.js]: It seems a little odd to ask `assetKit` and `metricsTopicKit` for storage nodes that were just provided _to_ them. Suggestion:
```diff
-  const metricsTopicKit = makeRecorderKit.async(
-    E(storageNode).makeChildNode('metrics'),
-  );
+  const metricsStorageNode = E(storageNode).makeChildNode('metrics');
+  const metricsTopicKit = makeRecorderKit.async(metricsStorageNode);
   …
   const topics = harden({
     asset: makePublicTopic(
       'State of the assets managed',
       assetKit.subscriber,
-      E(assetKit.recorder).getStorageNode(),
+      storageNode,
     ),
     metrics: makePublicTopic(
       'Vault Factory metrics',
       metricsTopicKit.subscriber,
-      E(metricsTopicKit.recorder).getStorageNode(),
+      metricsStorageNode,
     ),
   });
```

File: packages/zoe/src/contractSupport/recorder.js [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * @param {import('@agoric/zoe').Baggage} baggage
+ * @param {ERef<Marshaller>} marshaller
+ * @param {string} [kindName]
+ */
+export const prepareRecorder = (
+  baggage,
+  marshaller,
+  kindName = 'Chain Storage Publisher',
+) => {
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js]: When a kind name argument exists for a prepare function, don't we conventionally require it and place it immediately after the baggage argument?
```suggestion
export const prepareRecorder = (baggage, kindName, marshaller) => {
```
```suggestion
export const prepareRecorder = (baggage, marshaller) => {
```

**Review by gibson042 — CHANGES_REQUESTED** [review]
The Recorder name is great and seems to be defined in a good place, although I have some quibbles with its method names and with `makeRecorderKit.async`.

Regardless, the overall structure of defining and propagating to deeper layers a high-level function that accepts a storage node and leverages an argument-free `makeDurablePublishKit()` to return a kit containing a publisher-shaped facet that sends sequential values to both that node and to the other (subcriber-shaped) facet seems correct to me.

File: packages/zoe/src/contractSupport/recorder.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+ * - method names more aligned with [async iterators](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-iterator-interface)
+ * - omits fail()
+ * - adds getStorageNode() from its durable state
+ *
+ * Other names such as StoredPublisher or ChainStoragePublisher were considered, but found to be sometimes confused with *durability*, another trait of this class.
+ */
+
+/**
+ * @template T
+ * @typedef {{ getStorageNode: () => StorageNode, next(value: T): Promise<void>, return(value: T): Promise<void> }} Recorder
```
Comment (mhofman) [inline packages/zoe/src/contractSupport/recorder.js]: Thinking about it more, the fact these do not fulfill with `IteratorResult`s makes me thing we should pick other names than re-using iterator methods.

File: packages/zoe/src/contractSupport/recorder.js [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    {
+      getStorageNode() {
+        return this.state.storageNode;
+      },
+      async next(value) {
+        const { publisher, storageNode } = this.state;
+        const encoded = await E(marshaller).serialize(value);
+        const serialized = JSON.stringify(encoded);
+        await E(storageNode).setValue(serialized);
+        return publisher.publish(value);
```
Comment (mhofman) [inline packages/zoe/src/contractSupport/recorder.js]: To be sure, we do want to delay publishing the value and for a failure in writing to `storageNode` to prevent that publishing? That is a behavior change from today's implementation.

File: packages/inter-protocol/src/vaultFactory/vault.js [hunk 12]
```diff
@@ -873,7 +873,7 @@ export const prepareVault = (baggage, marshaller, zcf) => {
             phase,
           } = state;
           if (outerUpdater) {
-            outerUpdater.finish(helper.getStateSnapshot(Phase.TRANSFER));
+            outerUpdater.return(helper.getStateSnapshot(Phase.TRANSFER));
```
Comment (mhofman) [inline packages/inter-protocol/src/vaultFactory/vault.js]: `void` missing?

**Review by mhofman — COMMENTED** [review]
Couple small observations. I also share some of Richard's uneasiness about have sync and async versions of makers.

I actually keep wondering what in the upstream execution provides promises but expects to synchronously receive objects back. Is there really no way for these promises to be awaited before invoking the makers?

File: packages/inter-protocol/src/vaultFactory/vaultHolder.js (hunk 3 above)
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vaultHolder.js:31]: Oh yeah I dropped the runtime use of `valueShape` in some iteration. It had been used in a `mustMatch` before recording. I'll restore it.

I also intend for the shape to specify the TS type so that all the methods that take or return `value` are typed. Maybe a `TypedPattern` for now that takes a generic for what the type is. I don't think I can genericize Pattern because it's `any`.

File: packages/zoe/src/contractSupport/recorder.js (hunk 6 above)
Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js]: Agree. To resolve in chat.

File: packages/inter-protocol/src/vaultFactory/vaultManager.js (hunk 8 above)
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vaultManager.js]: For now it does. I'm working towards more encapsulation because the storageNode is only the concern of the topics, not the contract. Even the child node making I want to have outside this Exo and into a "topics" maker. It'll make more sense when that's done but I don't know if I'll have time to include it in this PR. Is this a blocker?

File: packages/zoe/src/contractSupport/recorder.js (hunk 9 above)
Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js]: fine by me

File: packages/zoe/src/contractSupport/recorder.js (hunk 11 above)
Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js]: Good to call out. Yes, to ensure that the two data sources don't get out of sync. If a Subscriber got the value and chainStorage didn't, there'd be a discrepancy.

File: packages/inter-protocol/src/vaultFactory/vaultDirector.js (hunk 1 above)
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vaultDirector.js]: Addressing this and related comment from @mhofman,
> I actually keep wondering what in the upstream execution provides promises but expects to synchronously receive objects back. Is there really no way for these promises to be awaited before invoking the makers?

There are some ERef<StorageNode> that can be awaited, but this one is in a `prepareKind` which must be sync: https://github.com/Agoric/agoric-sdk/blob/79bd8b5baef7fe3a4139e3b6982e82715fe12a2b/packages/inter-protocol/src/vaultFactory/vaultDirector.js#L114-L117

Moving the await up to the caller is bad separation of concerns. I could defer the creation of the topics object (making it a promise) but I was trying to avoid `getPublicTopics()` becoming async too. I guess it's fine since it's a remote method anyway. I'll try that.

File: packages/inter-protocol/src/vaultFactory/vaultManager.js (hunk 8 above)
Comment (gibson042) [reply packages/inter-protocol/src/vaultFactory/vaultManager.js]: No, not a blocker.

File: packages/inter-protocol/src/vaultFactory/vaultDirector.js (hunk 1 above)
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vaultDirector.js]: I tried it. While having both sync and async versions adds complexity, the async contagion from trying to eschew EventualRecorder made the code even more complex and hard to follow. So I kept the async version around but separated it into a separate method as requested.

File: packages/zoe/src/contractSupport/recorder.js (hunk 5 above)
Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js:17]: no longer applicable.

File: packages/zoe/src/contractSupport/recorder.js (hunk 6 above)
Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js]: went with `write` / `writeFinal` since `recorder.record` was redundantly redundant and conflicts a bit with "record (n)".

File: packages/zoe/src/contractSupport/recorder.js (hunk 7 above)
Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js]: I went with something like this. Not "When" because it synchronously makes an EventualRecorderKit.

File: packages/inter-protocol/src/vaultFactory/vaultHolder.js (hunk 3 above)
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vaultHolder.js:31]: the `valueGuard` is in there now. I left them as `M.any()` because I'm not confident in how they'll be updated through upgrade, but I think this is the right API for specifying the value type

File: packages/inter-protocol/src/vaultFactory/vaultFactory.js (hunk 2 above)
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vaultFactory.js]: clarified

File: packages/zoe/src/contractSupport/recorder.js (hunk 9 above)
Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js]: oh I put kindName after marshaller because I wanted it to be optional. And the reason for that is I couldn't think of a reason the contract would be concerned with the kind name or want to have more than one per vat. Why is it important to have,
```js
const makeRecorder = prepareRecorder(baggage, 'Recorder', marshaller);
```
vs,
```js
const makeRecorder = prepareRecorder(baggage, marshaller);
```

Comment (gibson042) [reply packages/zoe/src/contractSupport/recorder.js]: I guess I'd turn that around and ask why it's important to have the kind name be optional rather than either required or (more likely) non-configurable? But I don't consider any of those approaches to be necessary, so consider this as fully optional feedback.

Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js]: non-configurable is ideal to me. sounds like you also think that would improve this PR so I'll go with that. (there's only ever one baggage and one storage marshaller so no reason to define a second kind within the vat)

File: packages/inter-protocol/src/vaultFactory/vaultManager.js [hunk 13]
```diff
@@ -1141,7 +1148,7 @@ export const prepareVaultManagerKit = (
 
     {
       finish: ({ state, facets: { helper } }) => {
-        assetPublisher.publish(
+        E(assetKit.recorderP).write(
```
Comment (mhofman) [inline packages/inter-protocol/src/vaultFactory/vaultManager.js]: `void` missing here?

File: packages/inter-protocol/src/vaultFactory/vaultManager.js [hunk 14]
```diff
@@ -174,20 +176,19 @@ export const prepareVaultManagerKit = (
...
     0n,
     factoryPowers.getGovernedParams().getChargingPeriod(),
   );
 
-  /** @type {PublishKit<AssetState>} */
-  const { publisher: assetPublisher, subscriber: assetSubscriber } =
-    makeVaultManagerPublishKit();
-  pipeTopicToStorage(assetSubscriber, storageNode, marshaller);
+  const assetKit = makeERecorderKit(
+    storageNode,
```
Comment (mhofman) [inline packages/inter-protocol/src/vaultFactory/vaultManager.js]: Isn't `storageNode` a presence now in this case, which would allow using `makeRecorderKit`?

**Review by mhofman — APPROVED** [review]
Out of curiosity, is there a reason `getPublicTopics` has to be synchronous? From what I gather, it's the reason that we need the `storageNode` to be an `ERef` in the topics, and that the EventualRecorder exist (I likely forgot other paths, but this seem the most obvious to me).

File: packages/inter-protocol/src/vaultFactory/vaultManager.js (hunk 14 above)
Comment (turadg) [reply packages/inter-protocol/src/vaultFactory/vaultManager.js]: Good catch! Will push that change

**Comment by turadg** [comment]
> Out of curiosity, is there a reason `getPublicTopics` has to be synchronous? From what I gather, it's the reason that we need the `storageNode` to be an `ERef` in the topics, and that the EventualRecorder exist (I likely forgot other paths, but this seem the most obvious to me).

When I made the change, to replace `EventualRecorderKit` with `Promise<RecorderKIt>`, I found the E chains it required to be more complex than this solution. https://github.com/Agoric/agoric-sdk/pull/7350#discussion_r1162091972

So the main motivation is legibility.

File: packages/zoe/src/contractSupport/recorder.js [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+
+/**
+ * Wrap a Publisher to record all the values to chain storage.
+ *
+ * @param {import('@agoric/zoe').Baggage} baggage
+ * @param {ERef<Marshaller>} marshaller
+ */
+export const prepareRecorder = (baggage, marshaller) => {
+  /**
+   */
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js]: Unnecessary empty JSDoc comment?
```suggestion
```

File: packages/zoe/src/contractSupport/recorder.js [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+        const { storagePath: heldPath } = this.state;
+        if (heldPath === undefined) {
+          const path = await E(this.state.storageNode).getPath();
+          this.state.storagePath = path;
+          return path;
+        } else {
+          // balance awaits
+          await null;
+          return heldPath;
+        }
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js]:
```suggestion
        // end synchronous prelude
        await null;
        if (heldPath !== undefined) {
          return heldPath;
        }
        const path = await E(this.state.storageNode).getPath();
        this.state.storagePath = path;
        return path;
```

File: packages/zoe/src/contractSupport/recorder.js [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+       */
+      async writeFinal(value) {
+        const { closed, publisher, storageNode, valueGuard } = this.state;
+        !closed || Fail`cannot write to closed recorder`;
+        mustMatch(value, valueGuard);
+        const encoded = await E(marshaller).serialize(value);
+        const serialized = JSON.stringify(encoded);
+        await E(storageNode).setValue(serialized);
+
+        // below here differs from writeFinal()
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js:120]:
```suggestion
        // below here differs from write()
```

File: packages/zoe/src/contractSupport/recorder.js [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+        return publisher.publish(value);
+      },
+      /**
+       * Like `write` but prevents future writes and terminates the Subscriber.
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js]:
```suggestion
       * Like `write` but prevents future writes and terminates the publisher.
```

File: packages/zoe/src/contractSupport/recorder.js [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+ * @typedef {TM extends TypedMatcher<infer T> ? T : never} MatchedType
+ */
+
+/**
+ * @template T
+ * @param {string} description
+ * @param {RecorderKit<T> | EventualRecorderKit<T>} recorderKit
+ * @returns {import('@agoric/notifier').PublicTopic<T>}
+ */
+export const makePublicTopic = (description, recorderKit) => {
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js]: Should this be defined somewhere else? It seems more like code that depends upon recorder.js than like code that is a part of it.

File: packages/inter-protocol/src/vaultFactory/vaultManager.js [hunk 20]
```diff
@@ -564,7 +559,7 @@ export const prepareVaultManagerKit = (
             totalShortfallReceived: state.totalShortfallReceived,
           });
 
-          metricsPublisher.publish(payload);
+          return E(metricsTopicKit.recorderP).write(payload);
```
Comment (gibson042) [inline packages/inter-protocol/src/vaultFactory/vaultManager.js:562]: I'm not convinced that a synchronous `assetKit` is beneficial, given the necessarily asynchronous `metricsTopicKit` (e.g., use of `makeERecorderKit` for both seems better to me). But I'm not so certain of this that I consider it a blocker, so I'll defer to your judgment on it.

File: packages/zoe/src/contractSupport/recorder.js [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+     * @param {StorageNode} storageNode
+     * @param {TypedMatcher<T>} [valueGuard]
+     */
+    (publisher, storageNode, valueGuard = M.any()) => {
+      return {
+        closed: false,
+        publisher,
+        storageNode,
+        storagePath: /** @type {string | undefined} */ (undefined),
+        valueGuard,
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js]:
```suggestion
     * @param {TypedMatcher<T>} [valueShape]
     */
    (publisher, storageNode, valueShape = M.any()) => {
      return {
        closed: false,
        publisher,
        storageNode,
        storagePath: /** @type {string | undefined} */ (undefined),
        valueShape,
```
(per https://github.com/Agoric/agoric-sdk/blob/f4a34abadce17db3f68d611419cb41bdc3f4f4d9/packages/store/src/types.js#L54-L60 and https://github.com/endojs/endo/blob/master/packages/patterns/README.md )

File: packages/inter-protocol/src/vaultFactory/vaultHolder.js (hunk 3 above)
Comment (gibson042) [reply packages/inter-protocol/src/vaultFactory/vaultHolder.js:31]: I guess I would suggest switching from tuple to record:
```suggestion
/** @type {{ [name: string]: { description: string, valueShape: Pattern } }} */
const PUBLIC_TOPICS = {
  vault: { description: 'Vault holder status', valueShape: M.any() },
```

File: packages/zoe/src/contractSupport/recorder.js [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+ * To be called at most once per baggage.
+ *
+ * `makeRecorderKit` is suitable for making a durable `RecorderKit` which can be held in Exo state.
+ * `makeERecorderKit` is for closures that must return a `subscriber` synchronously but can defer the `recorder`.
+ *
+ * @param {import('@agoric/vat-data').Baggage} baggage
+ * @param {ERef<Marshaller>} marshaller
+ */
+export const prepareRecorderKitMakers = (baggage, marshaller) => {
+  const makeDurablePublishKit = prepareDurablePublishKit(
```
Comment (gibson042) [inline packages/zoe/src/contractSupport/recorder.js:201]: It still feels overly obfuscated to have `prepareRecorderKitMakers` define two independent Kinds, and likely to bite us if a future change requires access to a non-Recorder publish kit. I really do think exposing and using the "define" pattern as the entry point is superior, but won't go further than a strong recommendation here.

**Review by gibson042 — APPROVED** [review]
There seems to be a lot of redundancy in packages/zoe/src/contractSupport/recorder.js, mostly to support synchronous handling that seems questionable, but it's not really hurting much.

Also, are these changes sufficiently covered by existing tests, or do we need more?

File: packages/zoe/src/contractSupport/recorder.js (hunk 19 above)
Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js]: fair. I'll move it.

File: packages/zoe/src/contractSupport/recorder.js (hunk 22 above)
Comment (turadg) [reply packages/zoe/src/contractSupport/recorder.js:201]: Agreed. It slowly dawned on me while refactoring this PR how easy it would be to make mistakes with the durable kind upgrades having the prepares hidden away
